pkg/build/generator/generator.go

```diff
@@ -93,6 +99,23 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 	if err != nil {
 		return nil, err
 	}
+
+	// Fetch the secrets associated with the BuilderServiceAccount.
+	// These secrets will be used for push/pull authentication
+	// FIXME: Add support for SourceSecret
+	if sa, err := g.ServiceAccounts.ServiceAccounts(bc.Namespace).Get(bootstrappolicy.BuilderServiceAccountName); err != nil {
+		glog.V(2).Infof("Unable to fetch builder service account for %s: %v", bc.Namespace, err)
+	} else {
+		for _, ref := range sa.Secrets {
+			secret, err := g.Secrets.Secrets(bc.Namespace).Get(ref.Name)
+			if err != nil {
+				glog.V(5).Infof("Unable to fetch builder secret %s/%s: %v", bc.Namespace, ref.Name, err)
+				continue
+			}
+			g.builderSecrets = append(g.builderSecrets, *secret)
```

Comment:
It's more a question of intent than success.  We create an ImagePullSecret and add it to both ImagePullSecrets and Secrets.  The user removes it from the ImagePullSecrets list.  Would that user still expect the secret to be used by the infrastructure for pulling an image?  Certainly it could be, by mounting the secret and using it manually (as this code does), but is that what is expected?  I'd argue no.

If only wanted possible, we'd just use the auto-mounted service account token and build the dockercfg inside the pod.  We have the power and that's guaranteed to always to work, but it's not what someone expects.